my/location/common.py

```diff
@@ -32,3 +32,48 @@ class Location(LocationProtocol):
     accuracy: Optional[float]
     elevation: Optional[float]
     datasource: Optional[str] = None  # which module provided this, useful for debugging
+
+
+def locations_to_gpx(locations: Iterable[LocationProtocol], buffer: TextIO) -> Iterator[Exception]:
+    """
+    Convert locations to a GPX file, printing to a buffer (an open file, io.StringIO, sys.stdout, etc)
+    """
+
+    try:
+        import gpxpy.gpx
+    except ImportError as ie:
+        from my.core.warnings import warn
+
+        warn("gpxpy not installed, cannot write to gpx. 'pip install gpxpy'")
+        raise ie
+
+    gpx = gpxpy.gpx.GPX()
+
+    # hmm -- would it be useful to allow the user to split this into tracks?, perhaps by date?
+
+    # Create first track in our GPX:
+    gpx_track = gpxpy.gpx.GPXTrack()
+    gpx.tracks.append(gpx_track)
+
+    # Create first segment in our GPX track:
+    gpx_segment = gpxpy.gpx.GPXTrackSegment()
+    gpx_track.segments.append(gpx_segment)
+
+
+    for location in locations:
+        try:
+            point = gpxpy.gpx.GPXTrackPoint(
+                latitude=location.lat,
+                longitude=location.lon,
+                elevation=location.elevation,
+                time=location.dt,
+                comment=location.datasource,
+            )
+        except AttributeError:
+            yield TypeError(
```

Comment:
there is also `@runtime_checkable` but never tried it myself, and it needs `typing_extensions`. So just in case you haven't seen it :)  https://mypy.readthedocs.io/en/stable/protocols.html#using-isinstance-with-protocols

Reply:
Ah yeah, I saw that but disregarded because of the version requirement. I guess this would work even if the user duck-typed some other location class :+1: